File: app/src/main/java/com/aliernfrog/LacMapTool/utils/FileUtil.java
package com.aliernfrog.LacMapTool.utils;

import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.RequiresApi;
import androidx.core.content.FileProvider;
import androidx.documentfile.provider.DocumentFile;

import com.aliernfrog.LacMapTool.R;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.InputStream;
import java.io.OutputStream;

public class FileUtil {
    public static String removeExtension(String path) {
        int extensionIndex = path.lastIndexOf(".");
        if (extensionIndex == -1) return path;

        return path.substring(0, extensionIndex);
    }

    public static void copyFile(String source, String destination) throws Exception {
        File src = new File(source);
        InputStream in;
        OutputStream out;
        if (src.isFile()) {
            in =  new FileInputStream(source);
            out = new FileOutputStream(destination);
            byte[] buffer = new byte[1024];
            int length;
            while ((length = in.read(buffer)) > 0) {
                out.write(buffer, 0, length);
            }
            in.close();
            out.close();
        } else {
            File[] files = src.listFiles();
            if (files != null) {
                for (File file : files) {
                    copyFile(file.getPath(), destination + "/" + file.getName());
                }
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    public static void copyFile(String source, DocumentFile destination, Context context) throws Exception {
        File src = new File(source);
        ContentResolver resolver = context.getContentResolver();
        InputStream in;
        OutputStream out;
        if (src.isFile()) {
            in = new FileInputStream(source);
            out = resolver.openOutputStream(destination.getUri());
            byte[] buffer = new byte[1024];
            int length;
            while ((length = in.read(buffer)) > 0) {
                out.write(buffer, 0, length);
            }
            in.close();
            out.close();
        } else {
            File[] files = src.listFiles();
            if (files != null) {
                for (File file : files) {
                    copyFile(file.getPath(), destination.createFile("", file.getName()), context);
                }
            }
        }
    }

    public static void copyFile(DocumentFile source, String destination, Context context) throws Exception {
        File dst = new File(destination);
        ContentResolver resolver = context.getContentResolver();
        InputStream in;
        OutputStream out;
        if (!source.isDirectory()) {
            in = resolver.openInputStream(source.getUri());
            out = new FileOutputStream(destination);
            byte[] buffer = new byte[1024];
            int length;
            while ((length = in.read(buffer)) > 0) {
                out.write(buffer, 0, length);
            }
            in.close();
            out.close();
        } else {
            if (!dst.exists()) dst.mkdirs();
            DocumentFile[] files = source.listFiles();
            for (DocumentFile file : files) {
                copyFile(file, destination + "/" + file.getName(), context);
            }
        }
    }

    public static String readFile(String source) throws Exception {
        BufferedReader reader = new BufferedReader(new FileReader(new File(source).getPath()));
        String _line;
        StringBuilder _full = new StringBuilder();
        while ((_line = reader.readLine()) != null) {
            if (_full.length() > 0) {
                _full.append("\n").append(_line);
            } else {
                _full.append(_line);
            }
        }
        reader.close();
        return _full.toString();
    }

    public static Intent shareFile(String source, String type, Context context) {
        File file = new File(source);
        Intent intent = null;
        if (file.exists()) {
            String packageName = context.getApplicationContext().getPackageName();
            Uri uri = FileProvider.getUriForFile(context, packageName+".provider", file);
            intent = new Intent(Intent.ACTION_SEND)
                    .setType(type)
                    .putExtra(Intent.EXTRA_STREAM, uri)
                    .putExtra(Intent.EXTRA_SUBJECT, R.string.info_sharing)
                    .putExtra(Intent.EXTRA_TEXT, R.string.info_sharing);
        }
        return intent;
    }

    public static void deleteDirectory(File directory) {
        if (directory.exists()) {
            File[] files = directory.listFiles();
            if (files != null) {
                for (File file : files) {
                    delete(file);
                }
            }
        }
        directory.delete();
    }

    public static void deleteDirectoryContent(File directory) {
        File[] files = directory.listFiles();
        if (files != null) {
            for (File file : files) {
                delete(file);
            }
        }
    }

    public static void delete(File file) {
        if (file.isDirectory()) {
            deleteDirectory(file);
        } else {
            file.delete();
        }
    }
}

File: app/src/main/java/com/aliernfrog/lactool/util/staticutil/FileUtil.kt
package com.aliernfrog.lactool.util.staticutil

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.provider.DocumentsContract
import android.text.format.DateUtils
import androidx.core.content.FileProvider
import com.aliernfrog.lactool.R
import com.aliernfrog.lactool.impl.FileWrapper
import com.aliernfrog.lactool.util.extension.toPath
import java.io.File
import java.io.OutputStream

class FileUtil {
    companion object {
        fun removeExtension(path: String): String {
            val extensionIndex = path.lastIndexOf(".")
            if (extensionIndex == -1) return path
            return path.substring(0, extensionIndex)
        }

        fun getFilePath(path: String): String {
            return if (path.startsWith("/")) path
            else Uri.parse(path).toPath()
        }

        fun getUriForPath(path: String): Uri {
            return DocumentsContract.buildDocumentUri(
                "com.android.externalstorage.documents",
                "primary:"+path.removePrefix("${Environment.getExternalStorageDirectory()}/")
            )
        }

        fun getTreeUriForPath(path: String): Uri {
            return DocumentsContract.buildTreeDocumentUri(
                "com.android.externalstorage.documents",
                "primary:"+path.removePrefix("${Environment.getExternalStorageDirectory()}/")
            )
        }

        fun lastModifiedFromLong(lastModified: Long?, context: Context): String {
            val lastModifiedTime = lastModified ?: System.currentTimeMillis()
            return DateUtils.getRelativeDateTimeString(
                /* c = */ context,
                /* time = */ lastModifiedTime,
                /* minResolution = */ DateUtils.SECOND_IN_MILLIS,
                /* transitionResolution = */ DateUtils.DAY_IN_MILLIS,
                /* flags = */ 0
            ).toString()
        }

        fun copyDirectory(source: File, target: File) {
            if (!target.isDirectory) target.mkdirs()
            source.listFiles()!!.forEach { file ->
                val targetFile = File("${target.absolutePath}/${file.name}")
                if (file.isDirectory) copyDirectory(file, targetFile)
                else file.inputStream().use { inputStream ->
                    targetFile.outputStream().use { outputStream ->
                        inputStream.copyTo(outputStream)
                    }
                }
            }
        }

        fun writeFile(outputStream: OutputStream, content: String) {
            val writer = outputStream.writer(Charsets.UTF_8)
            writer.write(content)
            writer.flush()
            writer.close()
        }

        fun openTextAsFile(text: String, context: Context) {
            val targetFile = File("${context.externalCacheDir}/shared/temptext.txt")
            targetFile.parentFile?.mkdirs()
            if (targetFile.isFile) targetFile.delete()
            val output = targetFile.outputStream()
            writeFile(output, text)
            val uri = FileProvider.getUriForFile(context, "${context.packageName}.provider", targetFile)
            val intent = Intent(Intent.ACTION_VIEW)
                .setDataAndType(uri, "text/plain")
                .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            context.startActivity(intent)
        }

        fun shareFiles(vararg files: FileWrapper, context: Context, title: String = context.getString(R.string.action_share)) {
            val isSingle = files.size <= 1
            val sharedFileUris = files.map {
                FileProvider.getUriForFile(context, "${context.packageName}.provider", moveToSharedCache(it, context))
            }
            val firstUri = sharedFileUris.first()
            val intent = Intent(
                if (sharedFileUris.size > 1) Intent.ACTION_SEND_MULTIPLE else Intent.ACTION_SEND
            )
                .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                .setDataAndType(firstUri, context.contentResolver.getType(firstUri))
            if (isSingle) intent.putExtra(Intent.EXTRA_STREAM, firstUri)
            else intent.putExtra(Intent.EXTRA_STREAM, ArrayList(sharedFileUris))
            context.startActivity(Intent.createChooser(intent, title))
        }

        private fun moveToSharedCache(file: FileWrapper, context: Context): File {
            val inputStream = file.inputStream(context)
            val targetFile = File("${context.externalCacheDir}/shared/${file.name}")
            targetFile.parentFile?.mkdirs()
            if (targetFile.isFile) targetFile.delete()
            val output = targetFile.outputStream()
            inputStream?.copyTo(output)
            inputStream?.close()
            output.close()
            return File(targetFile.absolutePath)
        }
    }
}